File: DashboardBrasileirãoOO/DasboardBrasileirãoOOpython/src/analise/analisador_campeonato.py
import pandas as pd
from typing import List, Dict
from modelo.time import Time
from modelo.partida import Partida

class AnalisadorCampeonato:
    TIME_REGIAO_MAP = {
        "Flamengo": "Sudeste", "Fluminense": "Sudeste", "Vasco da Gama": "Sudeste", "Botafogo": "Sudeste",
        "Corinthians": "Sudeste", "Palmeiras": "Sudeste", "São Paulo": "Sudeste", "Santos": "Sudeste",
        "Grêmio": "Sul", "Internacional": "Sul", "Athletico-PR": "Sul", "Coritiba": "Sul",
        "Bahia": "Nordeste", "Fortaleza": "Nordeste", "Atlético-MG": "Sudeste", "Cruzeiro": "Sudeste",
        "América-MG": "Sudeste", "Goiás": "Centro-Oeste", "Cuiabá": "Centro-Oeste", "Bragantino": "Sudeste",
        "Paraná": "Sul", "Joinville": "Sul", "Santa Cruz": "Nordeste", "CSA": "Nordeste",
        "Juventude": "Sul", "Figueirense": "Sul", "EC Vitória": "Nordeste", "Ponte Preta": "Sul",
        "Avaí": "Sudeste", "Atlético-GO": "Centro-Oeste", "Chapecoense": "Sul", "Ceará SC": "Nordeste",
        "Sport Recife": "Nordeste"
    }

    def __init__(self, partidas_df: pd.DataFrame):
        self.partidas_df = partidas_df
        self.estatisticas_times = self._processar_partidas()

    def _processar_partidas(self) -> Dict[str, Time]:
        estatisticas = {}
        for _, row in self.partidas_df.iterrows():
            time_casa_nome = row['time_casa'].strip()
            time_fora_nome = row['time_fora'].strip()

            if time_casa_nome not in estatisticas:
                estatisticas[time_casa_nome] = Time(time_casa_nome)
            if time_fora_nome not in estatisticas:
                estatisticas[time_fora_nome] = Time(time_fora_nome)

            time_casa = estatisticas[time_casa_nome]
            time_fora = estatisticas[time_fora_nome]

            if row['gols_casa'] > row['gols_fora']:
                time_casa.adicionar_vitoria(row['gols_casa'], row['gols_fora'])
                time_fora.adicionar_derrota(row['gols_fora'], row['gols_casa'])
            elif row['gols_fora'] > row['gols_casa']:
                time_fora.adicionar_vitoria(row['gols_fora'], row['gols_casa'])
                time_casa.adicionar_derrota(row['gols_casa'], row['gols_fora'])
            else:
                time_casa.adicionar_empate(row['gols_casa'])
                time_fora.adicionar_empate(row['gols_fora'])
        return estatisticas

    @staticmethod
    def get_regiao_do_time(nome_time: str) -> str:
        return AnalisadorCampeonato.TIME_REGIAO_MAP.get(nome_time, "Desconhecida")

    def get_tabela_classificacao(self, ordem_crescente: bool, regioes_selecionadas: List[str]) -> List[Time]:
        times_filtrados = [
            time for time in self.estatisticas_times.values()
            if not regioes_selecionadas or self.get_regiao_do_time(time.nome) in regioes_selecionadas
        ]
        
        times_filtrados.sort(key=lambda t: (t.pontos, t.vitorias, t.saldo_gols, t.gols_pro), reverse=not ordem_crescente)
        return times_filtrados

    def get_nomes_dos_times(self, regioes_selecionadas: List[str]) -> List[str]:
        return sorted([
            nome for nome in self.estatisticas_times.keys()
            if not regioes_selecionadas or self.get_regiao_do_time(nome) in regioes_selecionadas
        ])
        
    def get_partidas_por_time(self, nome_time: str) -> List[Partida]:
        partidas_do_time = self.partidas_df[
            (self.partidas_df['time_casa'].str.strip() == nome_time) | 
            (self.partidas_df['time_fora'].str.strip() == nome_time)
        ]
        return [Partida(**row) for _, row in partidas_do_time.iterrows()]

    def get_performance_ao_longo_das_rodadas(self, nomes_times: List[str]):
        series_por_time = {nome: {"x": [], "y": []} for nome in nomes_times}
        pontos_acumulados = {nome: 0 for nome in nomes_times}
        
        partidas_ordenadas = self.partidas_df.sort_values(by=['rodada'])

        for _, p in partidas_ordenadas.iterrows():
            time_casa = p['time_casa'].strip()
            time_fora = p['time_fora'].strip()

            pontos_ganhos_casa = 0
            pontos_ganhos_fora = 0
            
            if p['gols_casa'] > p['gols_fora']:
                pontos_ganhos_casa = 3
            elif p['gols_fora'] > p['gols_casa']:
                pontos_ganhos_fora = 3
            else:
                pontos_ganhos_casa = 1
                pontos_ganhos_fora = 1
            
            if time_casa in nomes_times:
                pontos_acumulados[time_casa] += pontos_ganhos_casa
                series_por_time[time_casa]["x"].append(p['rodada'])
                series_por_time[time_casa]["y"].append(pontos_acumulados[time_casa])
            
            if time_fora in nomes_times:
                pontos_acumulados[time_fora] += pontos_ganhos_fora
                series_por_time[time_fora]["x"].append(p['rodada'])
                series_por_time[time_fora]["y"].append(pontos_acumulados[time_fora])
                
        return series_por_time

    def get_distribuicao_total_gols(self, num_anos: int, regioes_selecionadas: List[str]):
        distribuicao = {"0-30": 0, "31-40": 0, "41-50": 0, "51-60": 0, "61+": 0}
        times_filtrados = [
            time for time in self.estatisticas_times.values()
            if not regioes_selecionadas or self.get_regiao_do_time(time.nome) in regioes_selecionadas
        ]

        for time in times_filtrados:
            gols_por_ano = time.gols_pro // num_anos
            if gols_por_ano <= 30: distribuicao["0-30"] += 1
            elif gols_por_ano <= 40: distribuicao["31-40"] += 1
            elif gols_por_ano <= 50: distribuicao["41-50"] += 1
            elif gols_por_ano <= 60: distribuicao["51-60"] += 1
            else: distribuicao["61+"] += 1
        return distribuicao

    def get_distribuicao_media_gols(self, num_anos: int, regioes_selecionadas: List[str]):
        distribuicao = {"0,0-0,3": 0, "0,3-0,6": 0, "0,6-0,9": 0, "0,9-1,2": 0, "1,2+": 0}
        times_filtrados = [
            time for time in self.estatisticas_times.values()
            if not regioes_selecionadas or self.get_regiao_do_time(time.nome) in regioes_selecionadas
        ]

        for time in times_filtrados:
            media_gols_time = time.gols_pro / (num_anos * 38)
            if media_gols_time <= 0.3: distribuicao["0,0-0,3"] += 1
            elif media_gols_time <= 0.6: distribuicao["0,3-0,6"] += 1
            elif media_gols_time <= 0.9: distribuicao["0,6-0,9"] += 1
            elif media_gols_time <= 1.2: distribuicao["0,9-1,2"] += 1
            else: distribuicao["1,2+"] += 1
        return distribuicao

    def get_percentual_invencibilidade(self, regioes_selecionadas: List[str]):
        invencibilidade = {}
        times_filtrados = [
            time for time in self.estatisticas_times.values()
            if not regioes_selecionadas or self.get_regiao_do_time(time.nome) in regioes_selecionadas
        ]
        for time in times_filtrados:
            if time.jogos > 0:
                percentual = ((time.vitorias + time.empates) / time.jogos) * 100
                invencibilidade[time.nome] = percentual
        return dict(sorted(invencibilidade.items(), key=lambda item: item[1], reverse=True)[:10])

    def get_estadios_com_mais_gols(self, regioes_selecionadas: List[str]):
        estadios = self.partidas_df[
            self.partidas_df.apply(
                lambda row: (self.get_regiao_do_time(row['time_casa']) in regioes_selecionadas) or
                            (self.get_regiao_do_time(row['time_fora']) in regioes_selecionadas),
                axis=1
            )
        ]
        estadios['gols_totais'] = estadios['gols_casa'] + estadios['gols_fora']
        estadios_agrupados = estadios.groupby('estadio')['gols_totais'].sum().nlargest(10)
        return estadios_agrupados.to_dict()

    def get_historico_titulos_rebaixamentos(self, regioes_selecionadas: List[str]):
        historico = {}
        partidas_por_ano = self.partidas_df.groupby('ano')
        for ano, df_ano in partidas_por_ano:
            analisador_do_ano = AnalisadorCampeonato(df_ano)
            classificacao_anual = analisador_do_ano.get_tabela_classificacao(False, regioes_selecionadas)
            
            if not classificacao_anual:
                continue
            
            campeao = classificacao_anual[0]
            if campeao.nome not in historico: historico[campeao.nome] = [0, 0]
            historico[campeao.nome][0] += 1
            
            total_times = len(classificacao_anual)
            if total_times >= 20:
                for i in range(4):
                    rebaixado = classificacao_anual[total_times - 1 - i]
                    if rebaixado.nome not in historico: historico[rebaixado.nome] = [0, 0]
                    historico[rebaixado.nome][1] += 1
        return historico

    @staticmethod
    def filtrar_por_regioes(partidas_df: pd.DataFrame, regioes: List[str]):
        if not regioes:
            return partidas_df
        
        return partidas_df[
            partidas_df.apply(
                lambda row: (AnalisadorCampeonato.get_regiao_do_time(row['time_casa']) in regioes) or
                            (AnalisadorCampeonato.get_regiao_do_time(row['time_fora']) in regioes),
                axis=1
            )
        ]

File: java/src/main/java/com/brasileirao/dashboard/analise/AnalisadorCampeonato.java
package com.brasileirao.dashboard.analise;

import com.brasileirao.dashboard.modelo.Partida;
import com.brasileirao.dashboard.modelo.Time;
import com.brasileirao.dashboard.ui.DashboardApp;
import javafx.scene.chart.XYChart;

import java.util.*;
import java.util.stream.Collectors;

public class AnalisadorCampeonato {

    private final List<Partida> partidas;
    private final Map<String, Time> estatisticasTimes;
    private final Comparator<Time> comparadorDecrescente;
    private final Comparator<Time> comparadorCrescente;

    private static final Map<String, String> TIME_REGIAO_MAP = Map.ofEntries(
            Map.entry("Flamengo", "Sudeste"), 
            Map.entry("Fluminense", "Sudeste"),
            Map.entry("Vasco da Gama", "Sudeste"), 
            Map.entry("Botafogo", "Sudeste"),
            Map.entry("Corinthians", "Sudeste"), 
            Map.entry("Palmeiras", "Sudeste"),
            Map.entry("São Paulo", "Sudeste"), 
            Map.entry("Santos", "Sudeste"),
            Map.entry("Grêmio", "Sul"), 
            Map.entry("Internacional", "Sul"),
            Map.entry("Athletico-PR", "Sul"), 
            Map.entry("Coritiba", "Sul"),
            Map.entry("Bahia", "Nordeste"), 
            Map.entry("Fortaleza", "Nordeste"),
            Map.entry("Atlético-MG", "Sudeste"), 
            Map.entry("Cruzeiro", "Sudeste"),
            Map.entry("América-MG", "Sudeste"), 
            Map.entry("Goiás", "Centro-Oeste"),
            Map.entry("Cuiabá", "Centro-Oeste"), 
            Map.entry("Bragantino", "Sudeste"),
            Map.entry("Paraná", "Sul"),
            Map.entry("Joinville", "Sul"),
            Map.entry("Santa Cruz", "Nordeste"),
            Map.entry("CSA", "Nordeste"),
            Map.entry("Juventude", "Sul"),
            Map.entry("Figueirense", "Sul"),
            Map.entry("EC Vitória", "Nordeste"),
            Map.entry("Ponte Preta", "Sul"),
            Map.entry("Avaí", "Sudeste"),
            Map.entry("Atlético-GO", "Centro-Oeste"),
            Map.entry("Chapecoense", "Sul"),
            Map.entry("Ceará SC", "Nordeste"),
            Map.entry("Sport Recife", "Nordeste")
    );

    public AnalisadorCampeonato(List<Partida> partidas) {
        this.partidas = partidas;
        this.estatisticasTimes = new HashMap<>();

        this.comparadorDecrescente = Comparator.comparingInt(Time::getPontos).reversed()
                .thenComparingInt(Time::getVitorias).reversed()
                .thenComparingInt(Time::getSaldoGols).reversed()
                .thenComparingInt(Time::getGolsMarc).reversed();
        this.comparadorCrescente = Comparator.comparingInt(Time::getPontos)
                .thenComparingInt(Time::getVitorias)
                .thenComparingInt(Time::getSaldoGols)
                .thenComparingInt(Time::getGolsMarc);
        processarPartidas();
    }

    private void processarPartidas() {
        //Adiciona uma verificação de segurança antes de processar os times 
        // para previnir um NullpointerException se o nome do time for null ou vazio
        for (Partida p : partidas) {
            if (p.getTimeCasa() != null && !p.getTimeCasa().isEmpty()) {
                estatisticasTimes.putIfAbsent(p.getTimeCasa(), new Time(p.getTimeCasa()));
            }
            if (p.getTimeFora() != null && !p.getTimeFora().isEmpty()) {
                estatisticasTimes.putIfAbsent(p.getTimeFora(), new Time(p.getTimeFora()));
            }
        }
        for (Partida p : partidas) {
            Time timeCasa = estatisticasTimes.get(p.getTimeCasa());
            Time timeFora = estatisticasTimes.get(p.getTimeFora());
            // ver se time fora e casa são válidos antes de adicionar 
            if (timeCasa != null && timeFora != null) {
                if (p.getGolsCasa() > p.getGolsFora()) {
                    timeCasa.adicionarVitoria(p.getGolsCasa(), p.getGolsFora());
                    timeFora.adicionarDerrota(p.getGolsFora(), p.getGolsCasa());
                } else if (p.getGolsFora() > p.getGolsCasa()) {
                    timeFora.adicionarVitoria(p.getGolsFora(), p.getGolsCasa());
                    timeCasa.adicionarDerrota(p.getGolsCasa(), p.getGolsFora());
                } else {
                    timeCasa.adicionarEmpate(p.getGolsCasa());
                    timeFora.adicionarEmpate(p.getGolsFora());
                }
            }
        }
    }

    public static String getRegiaoDoTime(String nomeTime) {
        return TIME_REGIAO_MAP.getOrDefault(nomeTime, "Desconhecida");
    }

    public List<Time> getTabelaClassificacao(boolean ordemCrescente, List<String> regioesSelecionadas) {
        return estatisticasTimes.values().stream()
                .filter(time -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(time.getNome())))
                .sorted(ordemCrescente ? comparadorCrescente : comparadorDecrescente)
                .collect(Collectors.toList());
    }

    public List<String> getNomesDosTimes(List<String> regioesSelecionadas) {
        return estatisticasTimes.keySet().stream()
                .filter(nome -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(nome)))
                .sorted().collect(Collectors.toList());
    }

    public List<Partida> getPartidasPorTime(String nomeTime) {
        return partidas.stream()
                .filter(p -> p.getTimeCasa().equals(nomeTime) || p.getTimeFora().equals(nomeTime))
                .collect(Collectors.toList());
    }
    
    public List<XYChart.Series<Number, Number>> getSeriesGolsMarcadosVsSofridos(List<String> regioesSelecionadas) {
        return estatisticasTimes.values().stream()
            .filter(time -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(time.getNome())))
            .map(time -> {
                XYChart.Series<Number, Number> series = new XYChart.Series<>();
                series.setName(time.getNome());
                series.getData().add(new XYChart.Data<>(time.getGolsMarc(), time.getGolsSofr()));
                return series;
            })
            .collect(Collectors.toList());
    }
    
    public Map<String, XYChart.Series<Number, Number>> getPerformanceAoLongoDasRodadas(List<String> nomesTimes) {
        Map<String, XYChart.Series<Number, Number>> seriesPorTime = new HashMap<>();
        Map<String, Integer> pontosAcumulados = new HashMap<>();
        
        for (String nomeTime : nomesTimes) {
            XYChart.Series<Number, Number> series = new XYChart.Series<>();
            series.setName(nomeTime);
            seriesPorTime.put(nomeTime, series);
            pontosAcumulados.put(nomeTime, 0);
            series.getData().add(new XYChart.Data<>(0, 0));
        }

        List<Partida> partidasOrdenadas = partidas.stream()
                .sorted(Comparator.comparingInt(Partida::getRodada))
                .collect(Collectors.toList());

        for (Partida p : partidasOrdenadas) {
            String timeCasa = p.getTimeCasa();
            String timeFora = p.getTimeFora();
            boolean timeCasaSelecionado = nomesTimes.contains(timeCasa);
            boolean timeForaSelecionado = nomesTimes.contains(timeFora);

            if (timeCasaSelecionado || timeForaSelecionado) {
                int pontosGanhosCasa = 0;
                int pontosGanhosFora = 0;

                if (p.getGolsCasa() > p.getGolsFora()) {
                    pontosGanhosCasa = 3;
                } else if (p.getGolsFora() > p.getGolsCasa()) {
                    pontosGanhosFora = 3;
                } else {
                    pontosGanhosCasa = 1;
                    pontosGanhosFora = 1;
                }
                
                if(timeCasaSelecionado) {
                    pontosAcumulados.merge(timeCasa, pontosGanhosCasa, Integer::sum);
                }
                if(timeForaSelecionado) {
                    pontosAcumulados.merge(timeFora, pontosGanhosFora, Integer::sum);
                }

                if(timeCasaSelecionado) {
                    seriesPorTime.get(timeCasa).getData().add(new XYChart.Data<>(p.getRodada(), pontosAcumulados.get(timeCasa)));
                }
                if(timeForaSelecionado) {
                    seriesPorTime.get(timeFora).getData().add(new XYChart.Data<>(p.getRodada(), pontosAcumulados.get(timeFora)));
                }
            }
        }
        return seriesPorTime;
    }

    public Map<String, Integer> getDistribuicaoTotalGols(int numAnos, List<String> regioesSelecionadas) {
        Map<String, Integer> distribuicao = new LinkedHashMap<>();
        distribuicao.put("0-30", 0);
        distribuicao.put("31-40", 0);
        distribuicao.put("41-50", 0);
        distribuicao.put("51-60", 0);
        distribuicao.put("61+", 0);

        List<Time> timesFiltrados = estatisticasTimes.values().stream()
                .filter(time -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(time.getNome())))
                .collect(Collectors.toList());
        
        for (Time time : timesFiltrados) {
            int gols = time.getGolsMarc();
            int golsPorAno = gols/numAnos;
            if (golsPorAno <= 30) distribuicao.merge("0-30", 1, Integer::sum);
            else if (golsPorAno <= 40) distribuicao.merge("31-40", 1, Integer::sum);
            else if (golsPorAno <= 50) distribuicao.merge("41-50", 1, Integer::sum);
            else if (golsPorAno <= 60) distribuicao.merge("51-60", 1, Integer::sum);
            else distribuicao.merge("61+", 1, Integer::sum);
        }
        return distribuicao;
    }

    public Map<String, Double> getDistribuicaoMediaGols(int numAnos, List<String> regioesSelecionadas){
        Map<String, Double> distribuicaoMedia = new LinkedHashMap<>();
        distribuicaoMedia.put("0,0-0,3", 0.0); 
        distribuicaoMedia.put("0,3-0,6", 0.0);
        distribuicaoMedia.put("0,6-0,9", 0.0);
        distribuicaoMedia.put("0,9-1,2", 0.0);
        distribuicaoMedia.put("1,2+", 0.0);

        List<Time> timesFiltrados = estatisticasTimes.values().stream()
                .filter(time -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(time.getNome())))
                .collect(Collectors.toList());



/*         double auxClasses = 0.0;
        for (Time time : timesFiltrados){                     // usando alguns princípios de estatística para divisão de classe, com essa form pois os nums são relativamente baixos
            int gols = time.getGolsMarc();                    // não foi feito dessa forma para o método e seu antecessor anterior pois fiz os teste e achei q não era preciso pela forma de suas distribuições
            double mediaGolsTime = (double) (gols / (numAnos * 38));
            if(auxClasses < mediaGolsTime) auxClasses = mediaGolsTime;
        } */
        
        
        for (Time time : timesFiltrados) {
            int gols = time.getGolsMarc();
            double mediaGolsTime = (double) gols / (numAnos * 38); // quando o time cai a média dele na série A acaba caindo (foi intencional)


            if (mediaGolsTime <= 0.3) distribuicaoMedia.merge("0,0-0,3", 1.0, Double::sum); 
            else if (mediaGolsTime <= 0.6) distribuicaoMedia.merge("0,3-0,6", 1.0, Double::sum);
            else if (mediaGolsTime <= 0.9) distribuicaoMedia.merge("0,6-0,9", 1.0, Double::sum);
            else if (mediaGolsTime <= 1.2) distribuicaoMedia.merge("0,9-1,2", 1.0, Double::sum);
            else distribuicaoMedia.merge("1,2+", 1.0, Double::sum);
        }
        return distribuicaoMedia;
    }

    public Time getEstatisticasTime(String nomeTime) {
        return estatisticasTimes.get(nomeTime);
    }


    public Map<String, Long> getEstadiosComMaisGols(List<String> regioesSelecionadas) {
        return partidas.stream()
            .filter(p -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(p.getTimeCasa())) || regioesSelecionadas.contains(getRegiaoDoTime(p.getTimeFora())))
            .collect(Collectors.groupingBy(
                Partida::getEstadio,
                Collectors.summingLong(p -> p.getGolsCasa() + p.getGolsFora())
            ))
            .entrySet().stream()
            .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
            .limit(10)
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                Map.Entry::getValue,
                (e1, e2) -> e1,
                LinkedHashMap::new
            ));
    }

    public Map<String, Double> getPercentualInvencibilidade(List<String> regioesSelecionadas) {
        return estatisticasTimes.values().stream()
            .filter(time -> time.getJogos() > 0)
            .filter(time -> regioesSelecionadas.isEmpty() || regioesSelecionadas.contains(getRegiaoDoTime(time.getNome())))
            .collect(Collectors.toMap(
                Time::getNome,
                time -> ((double) (time.getVitorias() + time.getEmpates()) / time.getJogos()) * 100
            ))
            .entrySet().stream()
            .sorted(Map.Entry.<String, Double>comparingByValue().reversed())
            .limit(10)
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                Map.Entry::getValue,
                (e1, e2) -> e1,
                LinkedHashMap::new
            ));
    }

    public Map<String, int[]> getHistorico(List<String> regioesSelecionadas) {
        Map<String, int[]> historico = new HashMap<>(); // [títulos, top 4] havia pensdo em usar isso depois para outro gráfico de títulos

        Map<Integer, List<Partida>> partidasPorAno = partidas.stream()
                .collect(Collectors.groupingBy(Partida::getAno));

        for (Map.Entry<Integer, List<Partida>> entry : partidasPorAno.entrySet()) {
            List<Partida> partidasDoAno = entry.getValue();
            AnalisadorCampeonato analisadorDoAno = new AnalisadorCampeonato(partidasDoAno);
            List<Time> classificacaoAnual = analisadorDoAno.getTabelaClassificacao(false, regioesSelecionadas);

            if (classificacaoAnual.isEmpty()) continue;

            Time campeao = classificacaoAnual.get(0);
            historico.computeIfAbsent(campeao.getNome(), k -> new int[2])[0]++;

            int totalTimes = classificacaoAnual.size();
            if (totalTimes >= 20) {
                for (int i = 0; i < 4; i++) {
                    Time top = classificacaoAnual.get(totalTimes - 1 - i);
                    historico.computeIfAbsent(top.getNome(), k -> new int[2])[1]++;
                }
            }
        }
        return historico;
    }

    public static List<Partida> filtrarPorRegioes(List<Partida> partidas, List<String> regioes) {
        if (regioes == null || regioes.isEmpty()) {
            return partidas;
        }
        return partidas.stream()
            .filter(p -> {
                String regiaoCasa = TIME_REGIAO_MAP.getOrDefault(p.getTimeCasa(), "Desconhecida");
                String regiaoFora = TIME_REGIAO_MAP.getOrDefault(p.getTimeFora(), "Desconhecida");
                return regioes.contains(regiaoCasa) || regioes.contains(regiaoFora);
            })
            .collect(Collectors.toList());
    }
}